compiler/lcalc/compile_without_exceptions.ml

```diff
@@ -0,0 +1,455 @@
+open Utils
+module D = Dcalc.Ast
+module A = Ast
+
+(** hoisting *)
+
+(** The main idea around this pass is to compile Dcalc to Lcalc without using [raise EmptyError] nor
+    [try _ with EmptyError -> _]. To do so, we use the same technique as in rust or erlang to handle
+    this kind of exceptions. Each [raise EmptyError] will be translated as [None] and each
+    [try e1 with EmtpyError -> e2] as [match e1 with | None -> e2 | Some x -> x].
+
+    When doing this naively, this requires to add matches and Some constructor everywhere. We apply
+    here an other technique where we generate what we call `cuts`. Cuts are expression whom could
+    minimally [raise EmptyError]. For instance
+    [let x = <e1, e2, ..., en| e_just :- e_cons> * 3 in x + 1], the sub-expression
+    [<e1, e2, ..., en| e_just :- e_cons>] can produce an empty error. So we make a cut with a new
+    variable [y] linked to the Dcalc expression [<e1, e2, ..., en| e_just :- e_cons>], and we return
+    as the translated expression [let x = y * 3 in x + 1].
+
+    The compilation of expressions is found in the functions [translate_and_cut ctx e] and
+    [translate_expr ctx e]. Every option-generating expresion when calling [translate_and_cut] will
+    be cutted and later handled by the [translate_expr] function. Every other cases is found in the
+    translate_and_cut function. *)
+
+type cuts = D.expr Pos.marked A.VarMap.t
+(** cuts *)
+
+let pp_var (fmt : Format.formatter) (n : _ Bindlib.var) =
+  Format.fprintf fmt "%s_%d" (Bindlib.name_of n) (Bindlib.hash_var n)
+
+type info = { expr : A.expr Pos.marked Bindlib.box; var : A.expr Bindlib.var; is_pure : bool }
+(** information about the Dcalc variable : what is the corresponding LCalc variable; an expression
+    build correctly using Bindlib, and a boolean indicating whenever the variable should be matched
+    (false) or not (true). *)
+
+type ctx = info D.VarMap.t
+(** information context about variables in the current scope *)
+
+let pp_info (fmt : Format.formatter) (info : info) =
```

Comment:
All the `pp_*` should be moved to `Dcalc.Print` or `Lcalc.Print`.

Reply:
The printer for var is moved to Dcalc.Print/Lcalc.Print. But those printers are about data structures specific to this transformation, hence should stay here.